### PR TITLE
Show actual Job URL on test report

### PR DESCRIPTION
**Pull request by fjetter**
By fetching the Jobs info we can show the URL pointing to the actual job (e.g. OSX noci1)

With this PR

https://github.com/dask/distributed/runs/7666177056?check_suite_focus=true

image

Without

https://github.com/dask/distributed/actions/runs/2794492651

image

It also reduces the size of the HTML page generated by shrinking the DF before rendering. The HTML grew to 10-20MB already

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Nice! cc @ian-r-rose if you get a chance to take a look

File: continuous_integration/scripts/test_report.py [hunk 1]
```diff
@@ -102,6 +102,44 @@ def maybe_get_next_page_path(response: requests.Response) -> str | None:
...
+            cache[url] = jobs
+
+    df_jobs = pandas.DataFrame.from_records(jobs)
+    extracted = df_jobs.name.str.extract(
+        r"\(([\w\-]+), (\d\.\d+),\s([\w|\s]+)\)"
+    ).dropna()
+    df_jobs["OS"] = extracted[0]
+    df_jobs["python_version"] = extracted[1]
+    df_jobs["partition"] = extracted[2]
+    # We later need to join on this. Somehow the job ID is not part of the workflow schema and we have no other way to join
```
Comment (ian-r-rose) [inline continuous_integration/scripts/test_report.py:131]: Yeah, I think the pain of this is why I didn't do it in the first place. I think the reason is that artifacts are organized on a per-workflow basis. More than one job could upload to the same artifact name, so you can't necessarily determine which job is responsible for a given artifact.

File: continuous_integration/scripts/test_report.py [hunk 2]
```diff
@@ -258,40 +309,35 @@ def download_and_parse_artifacts(
...
+            # Note: we assign a column with the workflow timestamp rather
+            # than the artifact timestamp so that artifacts triggered under
+            # the same workflow can be aligned according to the same trigger
+            # time.
+            job = jobs_df[jobs_df["suite_name"] == a["name"]]
+            df2 = df.assign(
+                name=a["name"],
+                suite=suite_from_name(a["name"]),
+                date=w["created_at"],
+                html_url=job["html_url"].unique()[0],
```
Comment (ian-r-rose) [inline continuous_integration/scripts/test_report.py:332]: Given the fragility of the construction of `"suite_name"`, I'd recommend making this robust to the case where `job` has no entries.

**Review by ian-r-rose — COMMENTED** [review]
Nice, thanks for this @fjetter! I think this is looking pretty good. I believe we were already pretty close to hitting the maximum requests-per-hour with this workflow, so we should probably keep an eye on this to make sure the extra requests to the jobs endpoint doesn't tip us over the edge

File: continuous_integration/scripts/test_report.py [hunk 3]
```diff
@@ -102,6 +102,44 @@ def maybe_get_next_page_path(response: requests.Response) -> str | None:
...
+        except KeyError:
+            jobs = []
+            while total_count is None or len(jobs) < total_count:
+                resp = get_from_github(
+                    workflow["jobs_url"], {"per_page": 100, "page": page}
+                ).json()
+                if total_count is None:
+                    total_count = resp["total_count"]
+                jobs.extend(resp["jobs"])
+                page += 1
```
[the remark this answers is not in the document]
Comment (fjetter) [reply continuous_integration/scripts/test_report.py]: I used your approach. I don't have a preference

File: continuous_integration/scripts/test_report.py (hunk 2 above)
Comment (fjetter) [reply continuous_integration/scripts/test_report.py]: I'll raise an assert if this is empty. Shouldn't happen.

**Comment by fjetter** [comment]
If we hit the API rate limiting I suggest we [cache](https://github.com/actions/cache) the shelve artifacts which buffer the API requests. Since in this PR we're shelving the raw XML response instead of the parsed dataframe we can do this safely without destroying forward/backwards compatibility of the script and it should be safe to reuse these file caches

File: continuous_integration/scripts/test_report.py (hunk 1 above)
Comment (fjetter) [reply continuous_integration/scripts/test_report.py:126]: let's see how this evolves. We're not changing this frequently and I think the benefit of having links directly is worth the hassle.
I can't remember how much time it took me to always look up the proper job the error was referring to..

I'm sure there is a way to directly correlate this but I didn't want to invest the time fully understanding their data schema

File: continuous_integration/scripts/test_report.py (hunk 2 above)
Comment (ian-r-rose) [reply continuous_integration/scripts/test_report.py]: It's the type of thing that can happen if the test matrix changes (which has happened in the past)

**Review by ian-r-rose — APPROVED** [review]
LGTM, thanks @fjetter

**Pull request merged**